Web in play whats new (#9139)
google/WebFundamentals: src/content/en/android/trusted-web-activity/whats-new/index.md

project_path: /web/android/_project.yaml
book_path: /web/android/_book.yaml
description: Learn about the new features for web applications that use Trusted Web Activity.

> {# wf_published_on: 2020-11-25 #}
> {# wf_updated_on: 2020-11-25 #}

Reviewer: ```suggestion
{# wf_published_on: 2020-12-02 #}
{# wf_updated_on: 2020-12-02 #}
```

Reviewer: If we don't set the date a little ahead, this article won't be first in the list when it comes out.

{% include "web/_shared/contributors/adrianajara.html" %}

Since [Trusted Web Activity][1] was introduced last year, the Chrome team continues to work on the
product, making it easier to use with Bubblewrap, adding new features like the upcoming Google Play
Billing integration, and enabling it to work on more platforms, like Chrome OS. This article will
summarize the latest and upcoming updates for Trusted Web Activity.

## New Bubblewrap and Trusted Web Activity features

[Bubblewrap][2] helps you create apps that launch your PWAs inside a Trusted Web Activity, without
requiring knowledge of platform-specific tooling.

### Simplified setup flow

Previously, using Bubblewrap required manually setting up the Java Development Kit and the Android
SDK, both of which are error prone. The tool now offers to automatically download the external
dependencies when run for the first time.

You can still choose to use an existing installation of the dependencies, if you prefer to do so,
and the new `doctor` command helps find issues and recommends fixes to the configuration, which can
now be updated from the command line using the `updateConfig` command.

> ### Improved Wizard

Reviewer: ```suggestion
### Improved wizard
```

When creating a project with `init`, Bubblewrap needs information to generate the Android app. The

> tool extracts values from the Web Manifest and provides defaults where possible.

Reviewer: Web App Manfest?

Reviewer: I should have caught this before. Is this the Web App Manifest or something particular to TWAs?

> tool extracts values from the Web Manifest and provides defaults where possible.

Author: ```suggestion
tool extracts values from the Web App Manifest and provides defaults where possible.
```

> You can change those values when creating a new project, but the meaning of each field was not

Reviewer: ```suggestion
You can change those values when creating a new project, but previously the meaning of each field was not
```

> clear. The initialization dialogs were rebuilt with better description and validation for each

Reviewer: ```suggestion
clear. The initialization dialogs were rebuilt with better descriptions and validation for each
```

input field.

### display: fullscreen and orientation support
In some cases, you may want your application to use as much of the screen as possible and, when

> building PWAs, this is implemented by setting the `display` field from the Web Manifest to

Reviewer: ```suggestion
building PWAs, this is implemented by setting the `display` field from the Web App Manifest to
```

`fullscreen`.

> When Bubblewrap detects the fullscreen option in the Web Manifest, it will configure the Android

Reviewer: ```suggestion
When Bubblewrap detects the fullscreen option in the Web App Manifest, it will configure the Android
```

application to also launch in full screen, or [immersive mode][3], in Android specific terms.

> The `orientation` field from the Web Manifest defines whether the application should be started in

Reviewer: ```suggestion
The `orientation` field from the Web App Manifest defines whether the application should be started in
```

portrait mode, landscape mode, or in the orientation the device is currently using. Bubblewrap now

> reads the Web Manifest field and uses it as a default when creating the Android app.

Reviewer: ```suggestion
reads the Web App Manifest field and uses it as a default when creating the Android app.
```

> Both configurations can be customized as part of the `bubblewrap init` flow.

Reviewer: ```suggestion
You can customize both configurations can be customized as part of the `bubblewrap init` flow.
```

Reviewer: Changing passive to active voice. It's easier on the reader.

### AppBundles Output

[App Bundles][4] is a publishing format for apps that delegates the final APK generation and
signing to Play. In practice, this enables smaller files to be served to users when downloading the
app from the store.

Bubblewrap now packages the application as an App Bundle, in a file called

> `app-release-bundle.aab`. You should prefer this format, when publishing apps to the Play Store,

Reviewer: ```suggestion
`app-release-bundle.aab`. You should prefer this format when publishing apps to the Play Store
```

generated by previous versions. To take advantage of the smaller binaries, you only need to update
your app using the latest version of Bubblewrap.

### How to update an existing app
An application generated by Bubblewrap is composed of a web application and a lightweight Android
wrapper that opens the PWA. Even though the PWA opened inside a Trusted Web Activity follows the
same update cycles as any web app, the native wrapper can and should be updated.

You should update your app to ensure it is using the latest version of the wrapper, with the latest
bug fixes and features. With the latest version of Bubblewrap installed, the `update` command will
apply the latest version of the wrapper to an existing project:

```shell
npm update -g @bubblewrap/cli
bubblewrap update
bubblewrap build
```

Another reason to update those applications is ensuring that changes to the Web Manifest are
applied to the application. Use the new `merge` command for that:

```
bubblewrap merge
bubblewrap update
bubblewrap build
```

## Updates to the Quality Criteria
Chrome 86 introduced changes to the Trusted Web Activity Quality Criteria, which are explained in

> full in this [blogpost][7].

Reviewer: ```suggestion
full in [Changes to quality criteria for PWAs using Trusted Web Activity][7].
```

A quick summary is that you should ensure your applications handle the following scenarios to
prevent them from crashing:

 - Failure to verify digital asset links at application launch
 - Failure to return HTTP 200 for an offline network resource request

>  - An HTTP 404 or 5xx error in the application.

Reviewer: ```suggestion
 - Return of an HTTP 404 or 5xx error in the application.
```

Reviewer: Lists should always use parallel grammatical structures.

  event.respondWith((async () => {
    try {
      return await fetchAndHandleError(event.request);
    } catch {
      // Failed to load from the network. User is offline or the response
      // has a status code that triggers the Quality Criteria.
      // Try loading from cache.
      const cachedResponse = await caches.match(event.request);
      if (cachedResponse) {
        return cachedResponse;
      }
      // Response was not found on the cache. Send the error / offline
      // page. OFFLINE_PAGE should be pre-cached when the service worker
      // is activated.
      return await caches.match(OFFLINE_PAGE);
    }
  })());
});

async function fetchAndHandleError(request) {
  const cache = await caches.open(RUNTIME_CACHE);
  const response = await fetch(request);

  // Throw an error if the response returns one of the status
  // that trigger the Quality Criteria.
  if (response.status === 404 ||
      response.status >= 500 && response.status < 600) {
    throw new Error(`Server responded with status: ${response.status}`);
  }

  // Cache the response if the request is successful.
  cache.put(request, response.clone());
  return response;
}
```

> [Workbox][9] bakes best practices and removes boiler plate when using service workers.

Reviewer: ```suggestion
[Workbox][9] bakes in best practices and removes boilerplate when using service workers.
```

    this.checkTrustedWebActivityCrash(response);

    // If it's a good response, it can be used as-is.
    return response;
  }

  // This callback is new in Workbox v6, and is triggered whenever
  // an error (including a NetworkError) is thrown when a handler runs.
  handlerDidError(details) {
    let fallbackURL;
    switch (details.error.details.error.type) {
      case 'E_NOT_FOUND': fallbackURL = this.notFoundFallbackUrl; break;
      case 'E_SERVER_ERROR': fallbackURL = this.serverErrorFallbackUrl; break;
      default: fallbackURL = this.offlineFallbackUrl;
    }

    return caches.match(fallbackURL, {
      // Use ignoreSearch as a shortcut to work with precached URLs
      // that have _WB_REVISION parameters.
      ignoreSearch: true,
    });
  }
}
```

## Google Play Billing

Besides allowing your app to sell digital goods and subscriptions on the Play Store,
[Google Play Billing][10] offers tools for managing your catalog, prices and subscriptions, useful
reports, and a checkout flow powered by the Play Store that is already familiar to your users. It
is also a requirement for applications published on the Play Store that sell digital goods.

> Chrome 88 will launch with an Origin Trial on Android that enables the integration of

Reviewer: ```suggestion
Chrome 88 will launch with an origin trial on Android that enables the integration of
```

[Trusted Web Activities][1], the [Payment Request API][11] and the [Digital Goods API][12] to
implement purchase flows via Google Play Billing. We expect this Origin Trial to also be available

> for Chrome OS on version 89.

Reviewer: How long do you expect the OTs to run? It would be good to say.

Author: I checked and it's not defined yet.

Reviewer: Please submit an update as soon as you know.

**Important:** The Google Play Billing API has its own [terminology][13] and includes client and
backend components. This section covers only a small part of the API that is specific to using the
Digital Goods API and Trusted Web Activity. Make sure to read the
[Google Play Billing documentation][14] and understand its concepts before integrating it into a
production application.

### The basic flow

<figure class="attempt-left">
  <img src="/web/android/trusted-web-activity/whats-new/play-menu.jpg"
      alt="Play Console menu"/>
  <figcaption>

>     <b>Figure 1</b>. Play Console menu

Reviewer: ```suggestion
    <b>Figure 1.</b> Play Console menu
```

  </figcaption>
</figure>

To provide digital goods via the Play Store you’ll need to configure your catalog on the Play
Store, as well as connect the Play Store as a payment method from your PWA.

When you are ready to configure your catalog, start by finding the Products section in the left
hand side menu on the Play Console:

<div class="clearfix"></div>

Here you’ll find the option to view your existing in-app products and subscriptions and you’ll also
find the create button for adding new ones.

> <!-- Add Image here -->

Reviewer 2: Can you remove this?

<figure class="center">
  <img src="/web/android/trusted-web-activity/whats-new/in-app-products.jpg"
      alt="In-app Products"/>
  <figcaption>

>     <b>Figure 2</b>. In-app Products.

Reviewer: ```suggestion
    <b>Figure 2.</b> In-app Products.
```

  </figcaption>
</figure>

<figure class="attempt-right">
  <img src="/web/android/trusted-web-activity/whats-new/product-details.jpg"
      alt="Product Details"/>
  <figcaption>

>     <b>Figure 3</b>. Product Details.

Reviewer: ```suggestion
    <b>Figure 3.</b> Product Details.
```

  </figcaption>
</figure>

> To create a new in-app product you will need a product ID, name, description and a price. It’s

Reviewer: ```suggestion
To create a new in-app product you will need a product ID, name, description, and a price. It’s
```

important to create meaningful and easy to remember product IDs, you’ll need them later and the IDs
can’t be changed once created.

When creating subscriptions you will also have to specify a billing period. You have the option to
list your subscription benefits and add features like whether you have a free trial, an
introductory price, a grace period, and a resubscribe option.

After creating each product, make them available from your app by activating them.

If you prefer you can add your products via the [Play Developers API][15].

<div class="clearfix"></div>

Once your catalog is configured, the next step is to configure the checkout flow from the PWA. You

> will use a combination for the [Digital Goods API][12] and the [Payment Request API][11] to achieve

Reviewer: ```suggestion
will use a combination of the [Digital Goods API][12] and the [Payment Request API][11] to achieve
```

this.

#### Fetch a product price with the Digital Goods API:

When using Google Play Billing, you will want to ensure that the price displayed to users matches

> the price from the store listing. Manually keeping those prices in sync would be impossible, and

Reviewer: ```suggestion
the price from the store listing. Manually keeping those prices in sync would be impossible, so
```

the Digital Goods API provides a way for the web application to query the underlying payment
provider for prices:

```javascript
// The SKU for the product, as defined in the Play Store interface
async function populatePrice(sku) {
  try {
    // Check if the Digital Goods API is supported by the browser.
    if (window.getDigitalGoodsService) {
      // The Digital Goods API can be supported by other Payments provider.
      // In this case, we're retrieving the Google Play Billing provider.
      const service =
          await window.getDigitalGoodsService("https://play.google.com/billing");

      // Fetch product details using the `getDetails()` method.
      const details = await service.getDetails([sku]);

      if (details.length === 0) {
        console.log(`Could not get SKU: "${sku}".`);
        return false;
      }

      // The details will contain both the price and the currenncy.
      item = details[0];
      const value = item.price.value;
      const currency = item.price.currency;

      const formattedPrice = new Intl.NumberFormat(navigator.language, {

>         style: 'currency', currency: currency }).format(value);

Reviewer 2: We could maybe have a comment here explaining why we need `Intl.NumberFormat`. Partially it's good practice, partially it's essential because otherwise the numbers will be "7.50000000" or something like that (and this is *not* a bug on our end).

      // Display the price to the user.
      document.getElementById("price").innerHTML = formattedPrice;
    } else {
      console.error("Could not get price for SKU \"" + sku + "\".");
    }
  } catch (error) {
    console.log(error);
  }
  return false;
}
```

> You can detect support for the Digital Goods API by checking if `getDigitalGoodsService` is

Reviewer: ```suggestion
You can detect support for the Digital Goods API by checking if `getDigitalGoodsService()` is
```

available on the `window` object.

Then call `window.getDigitalGoodsService()` with the Google Play Billing identifier as a parameter.
This will return a service instance for Google Play Billing and other vendors can implement support
for the Digital Goods API and will have different identifiers.

Finally, call `getDetails()` on the reference to the Google Play Billing object passing the SKU for
the item as a parameter. The method will return a detail object containing both the price and the
currency for the item that can be displayed to the user.

#### Start the purchase flow:

The Payment Request API enables purchase flows on the web and is also used for the Google Play

> Billing integration. Check out this [blog post][16] to learn more if you are new to the Payment

Reviewer: ```suggestion
Billing integration. Check out this [How Payment Request API Works][16] to learn more if you are new to the Payment
```

Request API.

In order to use the API with Google Play Billing you will need to add a payment instrument which

> has a supportedMethod called `https://play.google.com/billing` and add the SKU as part of the data

Reviewer: ```suggestion
has a supported metod called `https://play.google.com/billing` and add the SKU as part of the data
```

for the instrument:

```javascript
const supportedInstruments = [{
  supportedMethods: "https://play.google.com/billing",
  data: {
    sku: sku
  }
}]
```

> Then, build a PaymentRequest object as usual and use the API as usual

Reviewer: ```suggestion
Then, build a `PaymentRequest` object as usual and use the API as usual
```

```javascript
const request = new PaymentRequest(supportedInstruments, details);
```

#### Acknowledge the purchase

Once the transaction is complete, you will need to use the Digital Goods API to acknowledge the

> payment. The response object from the PaymentRequest will contain a token you will use to

Reviewer: ```suggestion
payment. The response object from the `PaymentRequest` will contain a token you will use to
```

result, it’s up to you to associate the purchase to the user in your backend and ensure they have
access to the purchased items. When associating the purchase to a user, remember to save the
purchase token, as you may need it to verify if the purchase has been cancelled or refunded, or if
a subscription is still active. Check out the [Real Time Developer Notifications API][17] and the
[Google Play Developer API][15] as they provide endpoints for handling those cases in your backend.

#### Check for existing entitlements

A user may have redeemed a promo code or may have an existing subscription to your product. In
order to validate that the user has the appropriate entitlements, you can call the
`listPurchases()` command on the digital goods service. This will return all purchases that your
customer has made in your app. This would also be the place to acknowledge any unacknowledged
purchases to ensure that the user correctly redeems their entitlements.

```javascript
const purchases = await itemService.listPurchases();
for (p of purchases) {
  if (!p.acknowledged) {
    await itemService.acknowledge(p.purchaseToken, “onetime”);
  }
}
```

## Upload to the Chrome OS Play Store

> Trusted Web Activities are also available, since Chrome 85 in the ChromeOS Play Store. The process

Reviewer: ```suggestion
Trusted Web Activities are also available since Chrome 85 in the Chrome OS Play Store. The process
```

> to list your app on the store is the same for Chrome OS as it is for Android.

Reviewer: ```suggestion
to list your app in the store is the same for Chrome OS as it is for Android.
```

Once you’ve created your app bundle, the [Play Console][18] will guide you through the required
steps to list the app on the Play Store. In the Play Console documentation you can find help to
[create your app listing][19], manage your apk files and other settings, as well as instructions
for testing and [safely releasing your app][20].

To restrict your application to Chromebooks only, add the `--chromeosonly` flag when initializing
the application in Bubblewrap:

```shell
bubblewrap init --manifest="https://example.com/manifest.json" --chromeosonly
```

When building your application manually, without Bubblewrap, add a `uses-feature` flag to your
Android manifest:

```xml
<uses-feature  android:name="org.chromium.arc" android:required="true"/>
```

> If your listing is shared with an Android app, the ChromeOS only package version will always have

Reviewer: ```suggestion
If your listing is shared with an Android app, the Chrome OS only package version will always have
```

> to be higher than the Android app package version. You can set up the ChromeOS bundle version to a

Reviewer: ```suggestion
to be higher than the Android app package version. You can set up the Chrome OS bundle version to a
```
